### PR TITLE
Use a sliding window for deltas

### PR DESCRIPTION
**Pull request by davidebbo**
This is not quite ready, but is for initial discussion about #40.

The reasoning I made is that too much choice is bad, and that maybe we should only offer sliding deltas since plain deltas are so noisy. We could rename it to sliding deltas.

I made the number of days we look back configurable (in code only), as `deltaInterval`. So please experiment by changing this hard coded value. Setting it to 1 is the old behavior. I set it to 4 for now, which significantly smoothens the delta graphs (while keeping their overall shape).

But I'd like to discuss a small problem: because I take the average delta between the first and last day (of the interval), it would be a more proper estimate for the **middle** day. But right now, I use it as the value for the last day. So if the deltas are increasing a lot, we end up with a number that's a little low for the date it's assigned to (and vice versa if going down). Potential solution is to make the sliding window be **centered** around the current date, and I could try that (not a hard change). That still leaves some pain point at the end of the graph, since there is no 'future' data. And obviously, this problem gets worse with larger sliding windows. There are probably some crazy extrapolation algorithms to deal with that if we care.

### CODE REVIEW

File: index.html [hunk 1]
```diff
@@ -541,11 +541,14 @@ <h1>{{ chosen_domain == 'US' ? 'USA' :
   scale = scale || 'log10';
   if (chart) { chart.detach(); }
   if (stat == 'deltas') {
```
Comment (davidbau) [inline index.html:543]: maybe allow the stat option string to specify the window width.  E.g.,
`
if (stat.match(/^\d+day$/)) {
  const deltaInterval = parseInt(stat.match(/^(\d+)day$/)[0]);
`

File: index.html [hunk 2]
```diff
@@ -541,11 +541,14 @@ <h1>{{ chosen_domain == 'US' ? 'USA' :
...
+    // How many days back do we go to calculate the delta
+    const deltaInterval = 4;
     extract.series = _.map(extract.series, p => [p[0], _.map(p[1], (d, i) => {
       if (i == 0) { return null; }
+      var prevIndex = Math.max(0, i - deltaInterval);
```
Comment (davidbau) [inline index.html:548]: Instead of pretending days before day 0 are the same as the value on day 0, we should put in the actual value.
To do this properly, we really shouldn't hack the beginning like this when there exist data before day 0.

That means we probably should change the `apply_start_rule` function.  Proposal: that function identifies the threshold day for each series and then drops data before day 0.  Instead, it should pad the arrays (with null or undefined values) to align day 0 in the same place for each series without dropping any data, and then the object it should return should have a `day_zero: i` field that is used later to plot day zero at the start.

Then here we can say if prevIndex < 0, we can assume the value is 0.

**Comment by davidbau** [comment]
On whether to label the sliding window with the endpoint or the center, and whether to use a fancy weighting kernel for the window.

I like you code as-is.  Simpler-to-explain is better.  Backward-looking stats are easy to explain.

For this chart, I'd like to avoid getting into the future-looking business - that's the domain of a predictive model, and something I think we don't want to touch.  I think it's simpler to say "avg change over previous 5 days" rather than "avg change over 5 day window centered at the day".  (which includes a couple future days and starts inviting the question about what exactly we mean by putting a number on a day where the future was unknown at that time).

**Comment by davidbau** [comment]
>
> too much choice is bad
>
Also - agreed on this for the user interface.  In the code we can maintain
flexibility, but then just have one particular choice presented to the
user.

File: index.html (hunk 2 above)
Comment (davidebbo) [reply index.html:548]: > Instead of pretending days before day 0 are the same as the value on day 0

To be precise, that's not exactly what it's doing. Instead, it's simply reducing the time window. e.g. if the first two numbers are 3 and 5, it'll end up with 4. You make it sound like it would end up with the average of 3,3,3,...,3,5, which would be very skewed toward the first number. So you're right that it's ignoring real data that comes before the graphing start, but it's doing it in a more reasonable way :)

Alternate fix idea would be to take delta processing out of `plot_series` and move it before `apply_start_rule`, which would make things work naturally without adding extra logic. But that could be disrupting some other things?

Though I'm wondering how much this really matters. If our delta interval is small (like 4), and the graph is going over a large set of days (it usually is), I'm guessing the difference would be barely noticeable between 'reducing the interval size at the edge' (what I have now) and doing the more correct thing by padding. It's about questioning whether the added complexity of keeping track of the `day_zero` index is worthwhile.

Your call, happy to do it either way!

File: index.html (hunk 1 above)
Comment (davidebbo) [reply index.html:543]: That's only worthwhile if we let the user change the size. And if we do, maybe it would feel cleaner to do it via a separate `chosen_delta_interval` variable to avoid having to parse out the delta option string?

I think I'll leave that out initially, and we can revisit. As long as the code is flexible on the interval size, we keep our options open.

File: index.html (hunk 2 above)
Comment (davidbau) [reply index.html:548]: After playing with it a bit, I think it's better not to shrink the window, or else we get weird artifacts like the first few cases in california plotted like below, where as the window gets bigger, the number decreases.  I think it's best to just treat "days before the first case" as zero, and "days before the plotted start date" as having their actual value, even though not visible in the plot.

image

Ideally we should do it right even if you manually edit the URL to choose a very large window like 30day.  As you noted it requires doing things to avoid interference with other functions, so I've coded it up as a change on top of your pull request.  Will merge soon.

**Comment by davidebbo** [comment]
Thanks for completing it. Played with it and behavior is great!
